improvements for contributors (#1432)
kubernetes-sigs/kubebuilder: CONTRIBUTING.md

# Contributing guidelines

> This document has the guidelines and the required information for who would like to contribute with. Currently, Kubebuilder officially support OSX and Linux platforms then, you may find issues to execute the following commands if you are using an Windows SO. 

Reviewer: ```suggestion
This document has the guidelines and the required information for who would like to contribute with. Currently, Kubebuilder officially support OSX and Linux platforms, you may find issues to execute the following commands if you are using Windows SO. 
```

> This document has the guidelines and the required information for who would like to contribute with it. Currently, Kubebuilder officially supports OSX and Linux platforms then, you may find issues to execute the following commands if you are using Windows SO.

Reviewer: ```suggestion
This document has the guidelines and the required information for who would like to contribute to this project. Currently, Kubebuilder officially supports OSX and Linux platforms, Windows is still not fully supported so you may find issues to execute some of the following commands under this OS.
```

> This document has the guidelines and the required information for who would like to contribute to this project. Currently, Kubebuilder officially supports OSX and Linux platforms, Windows is still not fully supported so you may find issues to execute some of the following commands under this OS.

Reviewer 2: I don't know that the first sentence adds anything that the reader doesn't already know (ie. I'm looking at the contributing guidelines because I want information about contributing to the project). The second sentence I would expect to find in the main project README, not here.

Reviewer: A small descrition as the first sentence is ok from my point of view. About the second sentence, I definetely agree that it shoul me in the README, but it could also be here.

Author: The first one clarifies the purpose of the doc and for who it is important. 
Moving the second one to the README.

Reviewer 2: I chose my words poorly. The first sentence, to me at least, is redundant to what the document _is_.

By this I mean a "Contributing guidelines" document is exactly that, a document with guidelines and other useful information for those who want to contribute to the project. From my perspective it would be like starting the `README` with a sentence saying "This is a document about xyz project".

It doesn't hurt to have it so consider it a nit. I was just making the point that it was not a substantive change.

Author: Done.

- Access to a Kubernetes v1.11.3+ cluster.

## Contributing steps

1. Submit an issue describing your proposed change to the repo in question.
1. The [repo owners](OWNERS) will respond to your issue promptly.
1. If your proposed change is accepted, and you haven't already done so, sign a Contributor License Agreement (see details above).
1. Fork the desired repo, develop and test your code changes.
1. Submit a pull request.

> ## How to build kubebuilder locally?

Reviewer: I would say that without the question mark looks better. It is a set of instruction, not a question.

Author: Makes sense. Thank you 👍

1. Ensure you activate module support before continue (`$ export GO111MODULE=on`)
1. Build the project by using the command `make build` 
1. Run the command `make install` to create a bin with the source code 

> 1. Then, run the tests by using the command `make go-test`. It will execute unit tests, and then you will be able to check your local configuration.

Reviewer: This should be an step of the next section instead of this, it has nothing to do with build and you have to do it always before a PR.

Author: This step is required to ensure the local setup.

Author: I changed as well to address your concern. 
Please, let me know wdyt now.

**NOTE** It will allow you to test the changes made since when you run `kubebuilder` commands, it will use the binary built from your source code. Run `which kubebuilder` to check its path.

> ## What to do before to push an pull request  

Reviewer 2: ```suggestion
## What to do before submitting a pull request
```

1. Build the project by using the command `make build` 
1. Run the command `make install` to create a bin with the source code 

> **NOTE** It will allow you to test the changes made since when you run `kubebuilder` commands, it will use the binary built from your source code. Run `which kubebuilder` to check its path.

Reviewer: I don't think this line adds anything to the previous content.

Author: These lines make clear for new users how to test the changes locally, that the make install will gen the bin based on the source code.

Author: It is not clear for new contributors. The idea is to make clear that it allow them test their changes locally

Reviewer: This section is called `How to build kubebuilder locally` and the next one was called `How to test kubebuilder locally`. With that info, do you still need to be told that if you build locally you will use the local binary?

Author: It is not clear for all dev. Many of them raising questions to understand that the make install will install the bin from the source code.

Author: @Adirio, 

However, I changed it to address your concerns and I am closing this one. 
Please, let me know wdyt now.

## How to verify your local setup

Run the tests by using the command `make go-test`. It will execute unit tests, and then you will be able to check your local configuration.

## What to before to push an pull request  

1. Run the script `make generate` to update/generate the mock data used in the e2e test in `$GOPATH/src/sigs.k8s.io/kubebuilder/testdata/` 
2. Run `make lint` to check your code.

> **NOTE** To run `make lint` is required to install `golangci-lint` locally. More info: https://github.com/golangci/golangci-lint#install

Reviewer: Instead of all the changes you did here, I would just add this as the third element in the old `How to test kubebuilder locally` and then add the table you created.

Author: IMO the other changes are required to clarify the information for new contributors.

Author: Usually, contributors do not know what they should do before doing a Pull Request. Also, note that the `make generate` is not a test and is something very specific to the project which requires further explanation. In this way, it cannot be in `How to test kubebuilder locally`.

Reviewer: `make generate` actually is a test, the golden one. It is very specific, yes, but it is still a test. By separating it out into two sections seems as if part of those tests do not need to be performed before pull requests.

Author: The `make generate` is NOT the test. It will generate the data. (golden.sh)
The  `make test-gen` will call the `golden_test.sh` which is a check/test.

Reviewer: True, `make test-gen` is the actual test. I still think adding this in three sections just makes the process harder to udnerstand to new contributors. Test setup (like `make generate`) and tests themselves (like `make test-gen`) are better in a single section IMO.

Author: It is not in 3 sections at all.  The `make generate` is just in the section `## What to before to push an pull request`.

Reviewer: Actually, they are 4 sections:
```
## How to build kubebuilder locally
## How to verify your local setup
## What to before to push an pull request
## How to test kubebuilder locally
```

Author: But are NOT the sections related to `make generate`. 
Each section is an diff topic and has a different purpose.

Author: PS.: I understand that you are against the sections. However, it is the previous standard of this doc as well as the other ones in the repo. I do not think that we should have 1 section with all.

Reviewer: There were two sections, four seems to be too much. They are small and the last 3 are rlated to the tests.

Author: Done changes to address your concern.  Let me know if you are ok with now.

Author: Marking it as closed.

1. Build the project by using the command `make build` 

> 1. Run the command `make install` to create a bin with the source code 

Reviewer: `make install` already calls `make build` so the first is completely redundant (it will actually compile it twice which is a waste of time).

Author: It is true.

**NOTE** In order to check the local environment run `make go-test`.

## What to do before submitting a pull request 

> 

Reviewer: Leave this, you need to run tests before submitting PRs.

Author: It is above:

> **NOTE** In order to check the local environment run `make go-test`.

Reviewer: Well, no test is mandatory then, as they are all run by the CI. It all depends on what you call mandatory.

Author: The make generate is not a test. It is mandatory since is a requirement keep the testdata updated based on the latest changes indeed for we are able to do the review.

1. Run the script `make generate` to update/generate the mock data used in the e2e test in `$GOPATH/src/sigs.k8s.io/kubebuilder/testdata/` 

> 2. Run `make lint` to check your code.

Reviewer: Leave all as ones as in the rest of the document, markdown will properly place the right number.

```suggestion
1. Run `make lint` to check your code.
```

Author: Move it for the NOTE since it is not really mandatory to.

**NOTE** To run `make lint` is required to install `golangci-lint` locally. More info: https://github.com/golangci/golangci-lint#install

**IMPORTANT:** The `make generate` is very helpful. By using it, you can check if good part of the commands still working successfully after the changes. Also, note that its usage is a pre-requirement to submit a PR.

Following the targets that can be used to test your changes locally.

|   Command	|   Description	|  Is called in the CI?  	|
|---	|---	|---	|

> | make go-test |  Runs go tests | no   	|

Reviewer: I think this test is being called inside `./test.sh` and `./check_testdata.sh` so technichally the CI is running it.

Author: No. It is not. See; 

https://github.com/kubernetes-sigs/kubebuilder/blob/1433321257750fde2bec5e34ab6c5b7115f5dafa/Makefile#L71-L73

Reviewer: Yes, it is. See:
https://github.com/kubernetes-sigs/kubebuilder/blob/7a259a399a5ef2e072f5b962a810968fd2cbdd4b/test.sh#L157

Author: ah I catcher what you mean now. it is over `Is called in the CI?`. 
Ok, but the target is not called used in the CI. So, I think it still false and true for the targets that are calling  ./test.sh and ./check_testdata.sh explicit.
